### ISSUE TITLE
[Form] Harden the form lifecycle to improve support for dynamic forms

### ISSUE THREAD

**Issue by webmozart** [symfony/symfony#8834]
Probably in Symfony 2.4 I'm going to make a few changes to the form architecture to enforce a proper form lifecycle. A BC layer will be provided until Symfony 3.0.

I'm writing this mainly as instructions to my future me, but anyone interested in the topic is invited to chime in :)
#### Summary
- `Form::setData()` will be deprecated in favor of `Form::initialize()`
- `Form::initialize()` can only ever be called once (usually automatically by the form system) guaranteeing the best possible performance
- `Form::initialize()` is guaranteed to be called on every element in the form tree (including those with "inherit_data" set), ensuring that all `*_SET_DATA` events are fired on time
- side effects from `Form::submit()`, `getData()`, `getNormData()` and `getViewData()` will be removed
- `DataMapperInterface` will be deprecated in favor of `DataMapper2Interface` (everybody cheer now ;) )
#### Question

After you read the text ( ;) ) please help me by answering the following question:

Are there use cases where calling `Form::setData()` is absolutely necessary and cannot be changed so that the default data is set in one of the other ways demonstrated below (last code sample)?
##### Motivation 1: Dynamic Forms

The motivation behind are dynamic forms (#5807). Dynamic forms rely on `setData()` being called in order to manipulate the form tree dynamically. This must be done before `get*Data()`, `submit()` or `createView()` are called, because these methods rely on the form tree being readily initialized.
##### Motivation 2: *_SET_DATA events and "inherit_data"

Currently, forms with the "inherit_data" option set cannot have *_SET_DATA event listeners, because `setData()` is never called on these forms (#8253, #8607, #8748).
##### Motivation 3: Performance

Depending on the number of nested forms and the number of event listeners, `setData()` can be a very expensive operation. The number of `setData()` calls were already reduced once in d4f4038f6daf7cf88ca7c7ab089473cce5ebf7d8 by introducing deferred data initialization. It is still possible to call `setData()` more than once though. Guaranteeing a single `setData()` call would also guarantee an optimal performance.
##### Motivation 4: Remove side effects

The result of deferred data initialization is that `getData()`, `getNormData()`, `getViewData()` and `submit()` now have side effects. I would like to remove those.
##### Motivation 5: Inline data mapping code

Currently, when writing a custom data mapper, most of the code from PropertyPathMapper has to be duplicated. In fact, the only implementation-dependent lines are

``` php
$form->setData($this->propertyAccessor->getValue($data, $propertyPath));
```

and

``` php
$this->propertyAccessor->setValue($data, $propertyPath, $form->getData());
```

Even worse, the other code in the data mappers is responsible for maintaining the correct lifecycle of the form. I would like to inline this critical code into `Form` and provide a new, more minimal data mapper interface.
#### Proposed Lifecycle

The proposed form lifecycle will consist of four phases, which partially overlap.

_Note in advance:_ I'm talking about the existing low-level API here. Don't confuse it with the high-level API that you typically use in your controllers (types, options, etc.)
##### Phase 1: Configuration

In this phase, form builders are created and configured.

The phase is complete when a form builder is turned into a form by calling `getForm()`. At this point, the default data and other configuration of the form are frozen.

``` php
$builder = $factory->createBuilder()
    ->setData('default')
    ->setDisabled(true)
    ->setCompound(false)
    // ...
```
##### Phase 2: Assembling

In this phase, the form tree is assembled by connecting the `Form` instances. Typically, this is done during `FormBuilder::getForm()`, but in dynamic forms you often need to call `Form::add()` and `Form::remove()` manually.

This phase is never really complete. The form tree can be changed in the later phases to allow a dynamic modification of the form tree.

``` php
// Configuration
$builder = $factory->createBuilder()
    ->add('name', 'text')
    ->add('email', 'email');

// Assembling
// Form instances are created, Form::add() is called
$form = $builder->getForm();

// Manual modifications
$form->add('submit', 'submit');
$form->remove('email');
```
##### Phase 3: Initialization

During initialization, the form tree is populated with default values. Initialization starts when `Form::initialize($defaultData = null)` is called on the root form. The call will cascade through the form tree.

After the call is complete, `initialize()` is guaranteed to have been called on every element in the form tree. If elements are added later on, `initialize()` will be called on them during `Form::add()`.

`initialize()` is very similar to the current `setData()`, i.e. the form's data is set and `*_SET_DATA` events are fired. Consequently, after the initialization of a form tree, all `*_SET_DATA` events are fired and their effects visible.

There are three major differences between `initialize()` and the current `setData()`:

First, `initialize()` must **only be called once**. Since `initialize()` cascades through the form tree, this implies that `initialize()` must only be called on the root form.

Second, the data argument in can be omitted. This happens mostly for the root form and when "mapped" is set to `false`. In this case, the default data will be retrieved from the form's configuration, which has been created in Phase 1.

``` php
$form = $factory->createBuilder('form', null, array('auto_initialize' => false))
    ->add('name', 'text')
    ->add('email', 'email')
    // passing the data as second argument to
    // FormFactory::createBuilder() or through the option
    // "data" is equivalent to the below statement
    ->setData(array('name' => 'foo', 'email' => 'bar'))
    ->getForm();

$form->initialize();
// sets the default data stored above

// implies $form->get('name')->initialize('foo')
// implies $form->get('email')->initialize('bar')
```

If the "auto_initialize" option is not set to `false`, as in this example, `initialize()` will be called automatically (this is already happening right now for your Symfony 2.3 forms).

Third and last, contrary to the current `setData()`, initialized forms cannot be added as children of other forms anymore.

``` php
$form = $factory->createForm('text', null, array('auto_initialize' => false));
$form->initialize('foo');

$otherForm->add($form); // boom

$form = $factory->createForm('text');

$otherForm->add($form); // boom, auto initialized
```
##### Phase 4: Submission

After a form tree is readily initialized and all `*_SET_DATA` events triggered, the form can be submitted. Forms must not be submitted before initialization is complete.

``` php
// Phase 1 and 2 inside the form type
// Phase 3 due to automatic initialization
$form = $this->createForm('my_type');

// Phase 4
$form->handleRequest($request);

if ($form->isSubmitted() && $form->isValid()) {
    // ...
}
```

The only difference will be that there will most likely be an "auto_submit" option with a default of `true`, sparing you of calling `handleRequest()` manually.

``` php
// Phase 1 and 2 inside the form type
// Phase 3 due to automatic initialization
// Phase 4 due to automatic submission
$form = $this->createForm('my_type');

if ($form->isSubmitted() && $form->isValid()) {
    // ...
}
```
#### Implementation

A few details on what changes are necessary to make the above happen.
##### Introduce new DataMapperInterface

A new data mapper interface will be introduced. Due to BC, it will have a very uninspring name:

``` php
interface DataMapper2Interface
{
    /**
     * Returns true if the data is considered empty. The
     * preconfigured default value will be used in this case.
     *
     * @param mixed $data
     *
     * @return Boolean
     */
    public function isEmpty($data);

    /**
     * Returns whether the data can be read. For example,
     * an ArrayAccessMapper will return true if and only if
     * $data is an array.
     *
     * @param mixed $data
     *
     * @return Boolean
     */
    public function isReadable($data);

    /**
     * Returns a textual representation of the types that
     * this mapper can read, which will be included in the
     * error message generated if isReadable() returns false.
     *
     * @return string
     */
    public function getReadableTypes();

    /**
     * Returns the mapped value from $data by accessing $path.
     * $path can be anything, depending on the implementation
     * (e.g. property path, XPath, array index etc.)
     *
     * @param mixed $data
     * @param mixed $path
     *
     * @return mixed The value read from $data
     */
    public function readValue($data, $path);

    /**
     * Writes a value into $data.
     *
     * @param mixed $data
     * @param mixed $path
     * @param mixed $value
     */
    public function writeValue($data, $path, $value)
}
```

This mapper can be used _alternatively_ to an implementation of the current `DataMapperInterface`. The other code contained in current `DataMapperInterface` implementation will be executed by `Form` directly (see Motivation 5).
##### Call initialize() on forms with "inherit_data" true

Once the data mapping code is inlined into `Form`, `initialize()` can also be called on forms with the "inherit_data" option set (see Motivation 2). This is currently impossible, because the current data mappers only have access to the parent's view data, but need to pass the parent's model data instead.
##### Remove side effects

The side effects from `getData()`, `getNormData()`,
`getViewData()` and `submit()` should be removed. Instead, exceptions should be thrown.

``` php
// old
if (!$this->defaultDataSet) {
    $this->setData($this->getConfig()->getData());
}

// new
if (!$this->defaultDataSet) {
    throw new RuntimeException('getData() must be called after initializing the form. Make sure "auto_initialize" is not disabled on the root form.');
}
```

This is possible, because `initialize()`, which is guaranteed to be called, also guarantees to fill the form with the appropriate default data.
##### Deprecate setData()

As last step, `setData()` will be deprecated. Consequently, you should only set default data during the configuration phase by one of the already existing means:

``` php
// passed to the factory
$form = $factory->create('form', $default);

// set on the builder
$form = $factory->createBuilder('form')
    ->setData($default)
    ->getForm();

// passed as option
$form = $factory->create('form', null, array('data' => $default));
$form->add('field', 'text', array('data' => $default));

// deprecated
$form->setData($default);
```

If you made it up to here, you have my full respect :) Please don' forget to answer my question above.

**Comment by ikodev** [comment]
Hi @webmozart,
I need a full validation skipping for a subform, however I discovered that it was/is (?) a functional limitation so I cannot use this code for my subform that need to use dynamic fields (select2 javascript library currently used).
```
$builder->addEventListener(FormEvents::POST_SUBMIT, function($event) {
            $event->stopPropagation();
        }, 900);
```
Is there any Enhancement since you wrote this issue ? I am on 2.6.11.

Thank you for your help.

**Comment by marfillaster** [comment]
@ikodev i'd add an autosubmit `SubmitType` field with validation_groups option set to false. This submit button will be hidden in view and `click` event should be triggered whenever a dynamic field is changed through JavaScript. Validation will be entirely skipped but make sure you handle the a false validity in the controller.

`if ($form->isSubmitted() && $form->isValid() && !$form->get('autosubmit')->isClicked()) {}`

**Comment by mhlavac** [comment]
Any news about this? I found that another possible solution is to use `property_path` with `getThis`, `setThis` methods. Could same approach be used to fix this issue internally?

``` php
class User
{
    public function getThis()
    {
        return $this;
    }

    public function setThis($this) {}
}

class UserType extends \Symfony\Component\Form\AbstractType
{
    public function buildForm(\Symfony\Component\Form\FormBuilderInterface $builder, array $options)
    {
        $builder->add('name', CompoundType::class, [
            'property_path' => 'this'
        ]);
    }
}
```

**Comment by winzou** [comment]
Hi guys, what happened to this issue? Is it solved in 3.0? Is there any workaround for previous versions? Thanks for letting us know what to do with it ;)

**Comment by emmanuelballery** [comment]
Regarding the "*_SET_DATA" event problem, I guess you can add a form extension in your project to dispatch the event precisely when the Form component does not. This solution works for your entire project and does not require you to tweak all your forms.

```php
namespace AppBundle\Form\Extension;

use Symfony\Component\Form\AbstractTypeExtension;
use Symfony\Component\Form\Extension\Core\Type\FormType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\Form\FormEvent;
use Symfony\Component\Form\FormEvents;
use Symfony\Component\Form\FormInterface;

class FormTypeExtension extends AbstractTypeExtension
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        // Only *_SET_DATA events are concerned
        foreach ([FormEvents::PRE_SET_DATA, FormEvents::POST_SET_DATA] as $eventName) {
            $builder->addEventListener($eventName, function (FormEvent $event) use ($eventName) {
                // Look for "inherit_data" child
                foreach ($event->getForm() as $child) {
                    if ($child instanceof FormInterface) {
                        if ($child->getConfig()->getInheritData()) {
                            // Create a new event and dispatch it to the child
                            $childEvent = new FormEvent($child, $event->getData());
                            $child->getConfig()->getEventDispatcher()->dispatch($eventName, $childEvent);
                        }
                    }
                }
            });
        }
    }

    public function getExtendedType()
    {
        return FormType::class;
    }
}
```

And register it like so :

```yml
services:
  app.form.extension.form_type_extension:
    class: AppBundle\Form\Extension\FormTypeExtension
    tags:
      - { name: 'form.type_extension', extended_type: 'Symfony\Component\Form\Extension\Core\Type\FormType' }
```

**Comment by VincentChalnot** [comment]
@emmanuelballery Your solution should be implemented natively in Symfony, plus there is absolutely no risk of backward compatibility break. You should create a PR.

**Comment by stof** [comment]
Well, a form using `inherit_data` would force to put some extra constraint on the listeners: the event listeners for `PRE_SET_DATA` are currently allowed to modify the data which is about to be set, but that should not be possible when inheriting.

**Comment by stof** [comment]
and this would force to implement some special dispatching anyway, as `Form::setData` is not called for such objects (and cannot, and should not), while it is the place where the event is dispatched currently.

**Comment by VincentChalnot** [comment]
I'm not sure to understand why it would be a bad thing to allow the ```Form::setData``` in case of a ```PRE_SET_DATA``` event on a form using inherit_data...
Because it could be called twice with the risk one event could overwrite the other?
But ok, I'm not discussing this.

Instead of saying that's is not possible to use ```*_SET_DATA``` events with ```inherit_data```, we could bubble up the event like in the proposed code but with a class extending ```FormEvent``` that would throw an exception in the ```setData```.
This would allow to use the event to do many other interesting things (like building the sub-form using the data) without breaking anything.

**Comment by emmanuelballery** [comment]
Hi! I was catching up on this and I noticed the [warning in the doc](https://symfony.com/doc/current/form/inherit_data_option.html) that I probably missed before.

![image](https://user-images.githubusercontent.com/1687685/46291976-1dc1c200-c590-11e8-9152-198400a86e98.png)

Event though it's probably a minor usage, I'm also interested in knowing why this is not supported and why forms with inherited data cannot modify the data.

> Instead of saying that's is not possible to use `*_SET_DATA` events with `inherit_data`, we could bubble up the event like in the proposed code but with a class extending `FormEvent` that would throw an exception in the `setData`. This would allow to use the event to do many other interesting things (like building the sub-form using the data) without breaking anything.

In my opinion, this solution from @VincentChalnot is neat and covers what's actually missing in 100% of my use cases :+1:

**Comment by stof** [comment]
@VincentChalnot `Form::setData` is not a dumb setter. It performs the processing of the data (data transformers & co) and sets it also in all sub-forms. So calling it on a Form inheriting data would break everything.

**Comment by emmanuelballery** [comment]
@stof we are talking about `FormEvent::setData` I guess, which is a dumb setter.

Using `setData` in inherited Form is indead [already prevented](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Form.php#L316).

**Comment by VincentChalnot** [comment]
Yes I was talking about ```FormEvent::setData```, and if it's already prevented then we can use the code provided by @emmanuelballery as-is.

**Comment by xabbuh** [comment]
I am closing this old issue as it never happens and I don't see it being realised anytime in the near future.

**Comment by RomanApunts** [comment]
> Regarding the "*_SET_DATA" event problem, I guess you can add a form extension in your project to dispatch the event precisely when the Form component does not. This solution works for your entire project and does not require you to tweak all your forms.
> 
> ```
> namespace AppBundle\Form\Extension;
> 
> use Symfony\Component\Form\AbstractTypeExtension;
> use Symfony\Component\Form\Extension\Core\Type\FormType;
> use Symfony\Component\Form\FormBuilderInterface;
> use Symfony\Component\Form\FormEvent;
> use Symfony\Component\Form\FormEvents;
> use Symfony\Component\Form\FormInterface;
> 
> class FormTypeExtension extends AbstractTypeExtension
> {
>     public function buildForm(FormBuilderInterface $builder, array $options)
>     {
>         // Only *_SET_DATA events are concerned
>         foreach ([FormEvents::PRE_SET_DATA, FormEvents::POST_SET_DATA] as $eventName) {
>             $builder->addEventListener($eventName, function (FormEvent $event) use ($eventName) {
>                 // Look for "inherit_data" child
>                 foreach ($event->getForm() as $child) {
>                     if ($child instanceof FormInterface) {
>                         if ($child->getConfig()->getInheritData()) {
>                             // Create a new event and dispatch it to the child
>                             $childEvent = new FormEvent($child, $event->getData());
>                             $child->getConfig()->getEventDispatcher()->dispatch($eventName, $childEvent);
>                         }
>                     }
>                 }
>             });
>         }
>     }
> 
>     public function getExtendedType()
>     {
>         return FormType::class;
>     }
> }
> ```
> 
> And register it like so :
> 
> ```yaml
> services:
>   app.form.extension.form_type_extension:
>     class: AppBundle\Form\Extension\FormTypeExtension
>     tags:
>       - { name: 'form.type_extension', extended_type: 'Symfony\Component\Form\Extension\Core\Type\FormType' }
> ```

That solution worked for me, just notice on current version of Symfony (5.0.8) this line  dont wor:

`$child->getConfig()->getEventDispatcher()->dispatch($eventName, $childEvent);`

The correct sintax:

`$child->getConfig()->getEventDispatcher()->dispatch($childEvent, $eventName);`

**Comment by not-a-lot** [comment]
> I am closing this old issue as it never happens and I don't see it being realised anytime in the near future.

I’m surprised you wrote that this issue never happens because I fairly quickly ran into it as a new user of Symfony when I wanted to use the `PRE_SET_DATA` event to set the default value of a ChoiceType field in a form that uses `inherit_data` :slightly_smiling_face:

**Comment by VincentChalnot** [comment]
I've been using the code provided by @emmanuelballery in production on several projects for a long time now and everything works smoothly. I think we can create a pull request with this.

**Comment by rwkt** [comment]
> I am closing this old issue as it never happens and I don't see it being realised anytime in the near future.
Just ran into this issue - it's little things like this that makes working with Symfony Forms a struggle (there's way too many unnecessary quirks / gotchas) when it ought to make development _easier_.

Hoping it can be officially supported!
